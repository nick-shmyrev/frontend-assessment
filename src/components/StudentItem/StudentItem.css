.student-item {
  border-bottom: 1px solid #ccc;
  padding: 20px 40px;
}

.flex {
  display: flex;
  align-items: flex-start;
}

.flex-between {
  justify-content: space-between;
}

.flex-container {
  display: flex;
  justify-content: flex-start;
  align-items: flex-start;
}

.student-pic {
  border-radius: 100%;
  border: 1px solid #ccc;
  height: 200px;
  width: 200px;
  margin-right: 40px;
}

.student-name {
  font-size: 42px;
  font-weight: 700;
  text-transform: uppercase;
  margin: 0 0 20px 0;
}

.student-details {
  padding-left: 30px;
}

.student-details p {
  margin: 0 0 5px;
  font-size: 18px;
  font-weight: 200;
}

.details-icon {
  cursor: pointer;
  width: 35px;
  height: 35px;
}

.details-icon div {
  width: 35px;
  height: 4px;
  background: #777;
  border-radius: 3px;
}

.details-icon:hover div {
  background: #000;
}
.details-icon div:first-child {
  transform: translateY(17px);
}
.details-icon div:last-child {
  transform: rotate(90deg) translateX(13px);
}
.details-icon-open div:last-child {
  display: none;
}

.test-details {
  margin-top: 20px;
}
