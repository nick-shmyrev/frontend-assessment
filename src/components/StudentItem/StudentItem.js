import React from 'react';
import { connect } from 'react-redux';
import { setStudentTag } from '../../actions/students-actions';

// Import styles
import style from './StudentItem.css';

export class StudentItem extends React.Component {
  state = {
    gradesShown: false,
  };
  
  getAverage = grades => (grades.reduce((a, c) => a + Number(c), 0) / grades.length);
  
  handleDetailsClick = () => {
    this.setState(prevState => {
      return { gradesShown: !prevState.gradesShown };
    });
  };
  
  renderGrades = () => (
    <div className={style['test-details']}>
      {this.props.grades.map((grade, index) => (
        <p key={index}>Test{index + 1}&nbsp;&nbsp;&nbsp;&nbsp;{grade}</p>
      ))}
    </div>
  );
  
  handleTagSubmit = (e) => {
    e.preventDefault();
    
    if (e.target['tagInput'].value.trim().length > 0) {
      this.props.setStudentTag(this.props.email, e.target['tagInput'].value.trim());
    }
    e.target['tagInput'].value = '';
  };
  
  renderTags = () => (
    <div>
      {this.props.tags.length > 0 && (
        this.props.tags.map(tag => <div key={tag} className={style.tag}>{tag}</div>)
      )}
      <form onSubmit={this.handleTagSubmit}>
        <input type="text" name="tagInput"/>
      </form>
    </div>
  );
  
  render() {
    return (
      <div className={style['student-item']}>
        <div className={[style.flex, style['flex-between']].join(' ')}>
          <div className={style['flex-container']}>
            <img
              src={this.props.pic}
              alt={this.props.firstName}
              className={style['student-pic']}
            />
  
            <div className="student-details">
              <h1 className={style['student-name']}>{this.props.firstName} {this.props.lastName}</h1>
    
              <div className={style['student-details']}>
                <p>Email: {this.props.email}</p>
                <p>Company: {this.props.company}</p>
                <p>Skill: {this.props.skill}</p>
                <p>Average: {this.getAverage(this.props.grades)}%</p>
                
                {this.state.gradesShown && (
                  <div>
                    {this.renderGrades()}
                    {this.renderTags()}
                  </div>
                )}
                
              </div>
            </div>
          </div>
          
          <div
            className={[style['details-icon'], this.state.gradesShown ? style['details-icon-open'] : null].join(' ')}
            onClick={this.handleDetailsClick}
          >
            <div/>
            <div/>
          </div>
    
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setStudentTag: (email, tag) => dispatch(setStudentTag(email, tag)),
});

export default connect(undefined, mapDispatchToProps)(StudentItem);
