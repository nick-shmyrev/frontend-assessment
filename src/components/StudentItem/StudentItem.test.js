import React from 'react';
import { shallow } from 'enzyme';
import { StudentItem } from './StudentItem';

describe('StudentItem component', () => {
  const student = {
    city: "FushÃ«-Muhurr",
    company: "Yadel",
    email: "iorton0@imdb.com",
    firstName: "Ingaberg",
    grades: [
      "78",
      "100",
      "92",
      "86",
      "89",
      "88",
      "91",
      "87"
    ],
    id: "1",
    lastName: "Orton",
    pic: "https://storage.googleapis.com/hatchways-app.appspot.com/assessments/data/frontend/images/voluptasdictablanditiis.jpg",
    skill: "Oracle",
    tags: ['tag 1'],
  };
  const setStudentTag = jest.fn();
  
  
  test('should render correctly', () => {
    const wrapper = shallow(<StudentItem {...student} setStudentTag={setStudentTag} />);
    
    expect(wrapper).toMatchSnapshot();
  });
  
  test('should set student tag correctly', () => {
    const value = 'tag';
    const wrapper = shallow(<StudentItem {...student} setStudentTag={setStudentTag} />);
    
    wrapper.find('.details-icon').simulate('click');
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {},
      target: {
        tagInput: { value },
      },
    });
    
    expect(setStudentTag).toHaveBeenLastCalledWith(student.email, value);
  });
});
