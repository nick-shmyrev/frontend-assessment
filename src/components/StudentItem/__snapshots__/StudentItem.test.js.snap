// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentItem component should render correctly 1`] = `
<div
  className="student-item"
>
  <div
    className="flex flex-between"
  >
    <div
      className="flex-container"
    >
      <img
        alt="Ingaberg"
        className="student-pic"
        src="https://storage.googleapis.com/hatchways-app.appspot.com/assessments/data/frontend/images/voluptasdictablanditiis.jpg"
      />
      <div
        className="student-details"
      >
        <h1
          className="student-name"
        >
          Ingaberg
           
          Orton
        </h1>
        <div
          className="student-details"
        >
          <p>
            Email: 
            iorton0@imdb.com
          </p>
          <p>
            Company: 
            Yadel
          </p>
          <p>
            Skill: 
            Oracle
          </p>
          <p>
            Average: 
            88.875
            %
          </p>
        </div>
      </div>
    </div>
    <div
      className="details-icon "
      onClick={[Function]}
    >
      <div />
      <div />
    </div>
  </div>
</div>
`;
